research/testing/openshift-zuul-test-alternatives.md

```diff
@@ -0,0 +1,360 @@
+FAILED tests_openshift/openshift_integration/test_pkgtool.py::Pkgtool::test_pkgtool_clone - requre.exceptions.ItemNotInStorage: Keys not in storage:/src/tests_openshift/openshift_integration/test_data/test...
+FAILED tests_openshift/openshift_integration/test_sandcastle.py::test_get_api_client - kubernetes.config.config_exception.ConfigException: Invalid kube-config file. No configuration found.
+========================== 2 failed, 172 passed, 3 skipped, 37 warnings in 70.40s (0:01:10) ==========================
+```
+
+I think the first one can be fixed improving this setup but not the second one:
+
+```python
+________________________________________________ test_get_api_client _________________________________________________
+
+    def test_get_api_client():
+        """let's make sure we can get k8s API client"""
+>       assert sandcastle.Sandcastle.get_api_client()
+
+tests_openshift/openshift_integration/test_sandcastle.py:9:
+_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
+/usr/local/lib/python3.9/site-packages/sandcastle/api.py:324: in get_api_client
+    load_kube_config(client_configuration=configuration)
+/usr/local/lib/python3.9/site-packages/kubernetes/config/kube_config.py:792: in load_kube_config
+    loader = _get_kube_config_loader(
+_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
+
+filename = '~/.kube/config', config_dict = None, persist_config = True
+kwargs = {'active_context': None, 'config_persister': <bound method KubeConfigMerger.save_changes of <kubernetes.config.kube_config.KubeConfigMerger object at 0x7ff3a1ea7d30>>}
+kcfg = <kubernetes.config.kube_config.KubeConfigMerger object at 0x7ff3a1ea7d30>
+
+    def _get_kube_config_loader(
+            filename=None,
+            config_dict=None,
+            persist_config=False,
+            **kwargs):
+        if config_dict is None:
+            kcfg = KubeConfigMerger(filename)
+            if persist_config and 'config_persister' not in kwargs:
+                kwargs['config_persister'] = kcfg.save_changes
+
+            if kcfg.config is None:
+>               raise ConfigException(
+                    'Invalid kube-config file. '
+                    'No configuration found.')
+E               kubernetes.config.config_exception.ConfigException: Invalid kube-config file. No configuration found.
+
+/usr/local/lib/python3.9/site-packages/kubernetes/config/kube_config.py:751: ConfigException
+```
+
+## Summary
+
+[`podman kube play`](https://docs.podman.io/en/v4.4/markdown/podman-kube-play.1.html) can not be used to test:
+
+- _sandcastle_; we need a k8s cluster to be able to use the `kubernates` library. We could deploy pods in the cluster using [`podman kube apply`](https://docs.podman.io/en/v4.4/markdown/podman-kube-apply.1.html) but still we need an up and running cluster.
```

Comment:
we could also implement 'container' plugin for sandcastle though that would be an overkill solution

Reply:
I don't think I know what you mean 😅 but I believe you 🙂

Reply:
Sorry, that wasn't fully correct what I wrote. In packit we have 2 command runners right now: local (via subprocess) or openshift (sandcastle).

We could create a 3rd one: execute a command inside a container (ideally in the sandcastle container image).